xl2times/transforms.py

```diff
@@ -2448,132 +2451,204 @@ def apply_transform_tables(
     model: TimesModel,
 ) -> dict[str, DataFrame]:
     """Include data from transformation tables."""
-    if Tag.tfm_upd in tables:
-        updates = tables[Tag.tfm_upd]
-        table = tables[Tag.fi_t]
-        new_tables = [table]
-        # Reset FI_T index so that queries can determine unique rows to update
-        tables[Tag.fi_t].reset_index(inplace=True, drop=True)
-
-        # TFM_UPD: expand wildcards in each row, query FI_T to find matching rows,
-        # evaluate the update formula, and add new rows to FI_T
-        # TODO perf: collect all updates and go through FI_T only once?
-        for _, row in tqdm(
-            updates.iterrows(),
-            total=len(updates),
-            desc=f"Applying transformations from {Tag.tfm_upd.value}",
-        ):
-            rows_to_update = query(
-                table,
-                row["process"],
-                row["commodity"],
-                row["attribute"],
-                row["region"],
-                row["year"],
-                row["val_cond"],
-            )
+    if Tag.tfm_comgrp in tables:
+        table = model.commodity_groups
+        updates = tables[Tag.tfm_comgrp].filter(table.columns, axis=1)
 
-            if not any(rows_to_update):
-                logger.info(f"A {Tag.tfm_upd.value} row generated no records.")
-                continue
+        commodity_groups = pd.concat([table, updates], ignore_index=True)
+        commodity_groups = commodity_groups.explode("commodity", ignore_index=True)
+        commodity_groups = commodity_groups.drop_duplicates()
+        commodity_groups.loc[commodity_groups["gmap"].isna(), ["gmap"]] = True
+        model.commodity_groups = commodity_groups.dropna()
 
-            new_rows = table.loc[rows_to_update].copy()
-            new_rows["source_filename"] = row["source_filename"]
-            eval_and_update(new_rows, rows_to_update, row["value"])
-            new_tables.append(new_rows)
-
-        # Add new rows to table
-        tables[Tag.fi_t] = pd.concat(new_tables, ignore_index=True)
-
-    if Tag.tfm_ins in tables:
-        table = tables[Tag.fi_t]
-        updates = tables[Tag.tfm_ins].filter(table.columns, axis=1)
-        tables[Tag.fi_t] = pd.concat([table, updates], ignore_index=True)
-
-    if Tag.tfm_dins in tables:
-        table = tables[Tag.fi_t]
-        updates = tables[Tag.tfm_dins].filter(table.columns, axis=1)
-        tables[Tag.fi_t] = pd.concat([table, updates], ignore_index=True)
-
-    if Tag.tfm_ins_txt in tables:
-        updates = tables[Tag.tfm_ins_txt]
-
-        # TFM_INS-TXT: expand row by wildcards, query FI_PROC/COMM for matching rows,
-        # evaluate the update formula, and inplace update the rows
-        for _, row in tqdm(
-            updates.iterrows(),
-            total=len(updates),
-            desc=f"Applying transformations from {Tag.tfm_ins_txt.value}",
+    for data_module in model.data_modules:
```

Comment:
I didn't review this loop in detail, because it looked like you just took the old code and moved it under the loop and changed the order in which different tags are processed? Was there any other logical change?

Reply:
No, it was just that. :-) Hopefully we could process the rows in parallel in one of the next PRs!